demos/projects/ST/b-l475e-iot01a/port/azure_iot_flash_platform.c

```diff
@@ -37,72 +43,200 @@ static AzureIoTResult_t prvBase64Decode( uint8_t * base64Encoded,
+        return eAzureIoTErrorFailed;
+    }
 
-    return eAzureIoTSuccess;
-}
+    mbedtls_md_context_t ctx;
+    mbedtls_md_type_t md_type = MBEDTLS_MD_SHA256;
 
-AzureIoTResult_t AzureIoTPlatform_EnableImage( AzureADUImage_t * const pxAduImage )
-{
-    (void)pxAduImage;
+    mbedtls_md_init( &ctx );
+    mbedtls_md_setup( &ctx, mbedtls_md_info_from_type( md_type ), 0 );
+    mbedtls_md_starts( &ctx );
+
+    AZLogInfo( ( "Starting the mbedtls calculation: image size %d\r\n", pxAduImage->ulImageFileSize ) );
+
+    for( size_t ulOffset = 0; ulOffset < pxAduImage->ulImageFileSize; ulOffset += sizeof( ucPartitionReadBuffer ) )
+    {
+        ulReadSize = pxAduImage->ulImageFileSize - ulOffset < sizeof( ucPartitionReadBuffer ) ? pxAduImage->ulImageFileSize - ulOffset : sizeof( ucPartitionReadBuffer );
 
-    // TODO: Fill in to program board to decide which memory bank to use on reboot.
+        memcpy(ucPartitionReadBuffer, (pxAduImage->xUpdatePartition + ulOffset), ulReadSize);
 
-    LogInfo( ( "AzureIoTPlatform_EnableImage()\r\n" ) );
+        mbedtls_md_update( &ctx, ( const unsigned char * ) ucPartitionReadBuffer, ulReadSize );
 
+        if( ( ulOffset % 65536 == 0 ) && ( ulOffset != 0 ) )
+        {
+            printf( "." );
+        }
+    }
+
+    printf( "\r\n" );
+
+    AZLogInfo( ( "Done\r\n" ) );
+
+    mbedtls_md_finish( &ctx, ucCalculatedHash );
+    mbedtls_md_free( &ctx );
+
+    if( memcmp( ucDecodedManifestHash, ucCalculatedHash, azureiotflashSHA_256_SIZE ) == 0 )
+    {
+        AZLogInfo( ( "SHAs match\r\n" ) );
+        xResult = eAzureIoTSuccess;
+    }
+    else
+    {
+        AZLogInfo( ( "SHAs do not match\r\n" ) );
+        AZLogInfo( ( "Wanted: " ) );
+
+        for( int i = 0; i < azureiotflashSHA_256_SIZE; ++i )
+        {
+            printf( "%x", ucDecodedManifestHash[ i ] );
+        }
+
+        printf( ( "\r\n" ) );
+        AZLogInfo( ( "Calculated: " ) );
+
+        for( int i = 0; i < azureiotflashSHA_256_SIZE; ++i )
+        {
+            printf( "%x", ucCalculatedHash[ i ] );
+        }
+
+        printf( ( "\r\n" ) );
+
+        xResult = eAzureIoTErrorFailed;
+    }
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTPlatform_EnableImage( AzureADUImage_t * const pxAduImage )
+{
+    FLASH_OBProgramInitTypeDef  xOptionBytes;
+
+    HAL_FLASH_Unlock();
+    HAL_FLASH_OB_Unlock();
+    HAL_FLASHEx_OBGetConfig(&xOptionBytes); // Get current optionbytes configuration
+    xOptionBytes.OptionType = OPTIONBYTE_USER;
+    xOptionBytes.USERType = OB_USER_BFB2;
+    // If BFB2 set, we will reset it
+    xOptionBytes.USERConfig =
+        ((xOptionBytes.USERConfig & OB_BFB2_ENABLE) == OB_BFB2_ENABLE)
+            ? OB_BFB2_DISABLE
+            : OB_BFB2_ENABLE;
+
+    HAL_FLASHEx_OBProgram(&xOptionBytes);
+
+    // sets options bits and restarts device.
+    // Also sets the book bank address to 0x08000000, which means we always write
+    // to 0x08080000
+    HAL_FLASH_OB_Launch();
```

Comment:
yeah, I wanted to chat a bit about this - ResetDevice never gets called because this reboots, but that also means that none of the things that run in sample_azure_iot_adu after AzureIoTPlatform_EnableImage get called either. So should the boot bank assignment happen in reset device instead and have this function be a no-op?

Reply:
I moved the OB_Launch to reset device, but ADU is showing the update as failed, and I think this might be because we haven't updated the version yet in code, but I'm not quite sure how to resolve this

Reply:
separated this out more cleanly and now I get that the update succeeds in ADU (the problem before was that we were running out of scratch buffer space, which would fatally end the update, and I think if we were in progress and then at the same version after a reboot, it gets marked as failed?)

Reply:
yes I'm almost positive if an update is sent as "in progress" as version 1.0 and then subsequently gets marked as "idle" for version 1.0, it's marked as a failure